light/txpool.go

```diff
@@ -224,30 +224,38 @@ func (pool *TxPool) reorgOnNewHead(ctx context.Context, newHeader *types.Header)
 	txc := make(txStateChanges)
 	oldh := pool.chain.GetHeaderByHash(pool.head)
 	newh := newHeader
-	// find common ancestor, create list of rolled back and new block hashes
 	var oldHashes, newHashes []common.Hash
-	for oldh.Hash() != newh.Hash() {
-		if oldh.Number.Uint64() >= newh.Number.Uint64() {
-			oldHashes = append(oldHashes, oldh.Hash())
-			oldh = pool.chain.GetHeader(oldh.ParentHash, oldh.Number.Uint64()-1)
-		}
-		if oldh.Number.Uint64() < newh.Number.Uint64() {
-			newHashes = append(newHashes, newh.Hash())
-			newh = pool.chain.GetHeader(newh.ParentHash, newh.Number.Uint64()-1)
-			if newh == nil {
-				// happens when CHT syncing, nothing to do
-				newh = oldh
+
+	if oldh != nil {
+		// find common ancestor, create list of rolled back and new block hashes
+		for oldh.Hash() != newh.Hash() {
+			if oldh.Number.Uint64() >= newh.Number.Uint64() {
+				oldHashes = append(oldHashes, oldh.Hash())
+				oldh = pool.chain.GetHeader(oldh.ParentHash, oldh.Number.Uint64()-1)
+			}
+			if oldh.Number.Uint64() < newh.Number.Uint64() {
+				newHashes = append(newHashes, newh.Hash())
+				newh = pool.chain.GetHeader(newh.ParentHash, newh.Number.Uint64()-1)
+				if newh == nil {
+					// happens when CHT syncing, nothing to do
+					newh = oldh
+				}
 			}
 		}
+		if oldh.Number.Uint64() < pool.clearIdx {
+			pool.clearIdx = oldh.Number.Uint64()
+		}
+		// roll back old blocks
+		for _, hash := range oldHashes {
+			pool.rollbackTxs(hash, txc)
+		}
+		pool.head = oldh.Hash()
 	}
-	if oldh.Number.Uint64() < pool.clearIdx {
-		pool.clearIdx = oldh.Number.Uint64()
-	}
-	// roll back old blocks
-	for _, hash := range oldHashes {
-		pool.rollbackTxs(hash, txc)
+
+	if newHeader == nil {
```

Comment:
This statement doesn't make much sense, especially in here.
Above, we assign `newHeader` to `newh`, and operate on `newh`. If `newHeader` was indeed `nil`, this would fail (except degenerated case when both `newh`/`newHeader` and `oldHeader` are `nil`).

If it actually can happen, or we want to be extra defensive, move the check above line 229.

Reply:
I was assuming `oldh` was nil...

Reply:
PTAL at changes thru https://github.com/etclabscore/core-geth/pull/54/commits/687e62601410c098332757dec8291df50daa4ffe.

Have added some test cases and simplified the checks.